### PR TITLE
Feature/adapter registration refactored

### PR DESCRIPTION
**Pull request by vladiIkor**
# Description

Please include a summary of the change and which issue (add a link) is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [ ] Test A
- [ ] Test B

**Test Configuration**:
* SIP Framework version(s):
* Other configuration:

# Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing (regression) unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by malte-westendorf-ikor — COMMENTED** [review]
I prefer this slimmer and more elegant approach.

File: sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/TelemetryDataCollector.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    this.adapterRouteEndpoint = adapterRouteEndpoint;
+    this.pathMappedEndpoints = pathMappedEndpoints;
+    this.healthEndpoint = healthEndpoint;
+  }
+
+  @Override
+  public TelemetryData collectData() {
+    telemetryData.setActuatorEndpoints(pathMappedEndpoints.getAllPaths());
+    telemetryData.setHealthStatus(this.healthEndpoint.health().getStatus());
+    telemetryData.setAdapterRoutes(this.adapterRouteEndpoint.routes());
```
Comment (malte-westendorf-ikor) [inline sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/TelemetryDataCollector.java]: This does not return the details of all routes.

Comment (vladiIkor) [reply sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/TelemetryDataCollector.java]: You are right, will be corrected.
Thanks!

File: sip-core/src/main/java/de/ikor/sip/foundation/core/SIPCoreConfiguration.java [hunk 2]
```diff
@@ -1,14 +1,29 @@
...
+      "de.ikor.sip.foundation.core.annotation",
+      "de.ikor.sip.foundation.core.proxies",
+      "de.ikor.sip.foundation.core.trace",
+      "de.ikor.sip.foundation.core.translate",
+      "de.ikor.sip.foundation.core.util"
+    })
+// @Import contains classes with conditional @ComponentScan. This way certain packages are loaded on
+// demand,
+// based on configuration.
+@Import({AdapterRegistration.class})
```
Comment (LetoBukarica) [inline sip-core/src/main/java/de/ikor/sip/foundation/core/SIPCoreConfiguration.java]: Probably a personal preference, but I would rather leave it as it was before (scan everything in core by default) and adding an exclude for premiumsupport package.

File: sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/AdapterRegistration.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  /**
+   * Is executed after the application has been started and its properties have been set, and
+   * scheduled for further invocations based on sip.core.backend-registration.interval property.
+   */
+  @Override
+  @Scheduled(fixedDelayString = "${sip.core.backend-registration.interval}")
+  public void afterPropertiesSet() {
+    this.registrationClient.registerAdapter();
+  }
```
Comment (LetoBukarica) [inline sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/AdapterRegistration.java]: We already discussed this. registerAdapter implies having a register/health in a loop/unregister flow. I don't actually know of a better name, just this one was kind of misleading. Maybe changing it to sendTelemetryData?

File: sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/SIPRegistrationProperties.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    this.checkOutUrl = this.platformUrl + this.checkOutPath;
+    this.checkInUrl = this.platformUrl + this.checkInPath;
+  }
+
+  /**
+   * This is a unique id that identifies an instance of an adapter. The value is generated and can
+   * not be overridden.
+   */
+  @Setter(AccessLevel.NONE)
+  private final UUID instanceId = UUID.randomUUID();
```
Comment (LetoBukarica) [inline sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/SIPRegistrationProperties.java:41]: This unique id looks like something that should be define in a broader scope and probably used on more than one place (actuator endpoints for example). It's kind of buried inside SIPRegistrationProperties. 

That being said, maybe there is a something Spring native that can give us the same result (like spring.application.instance_id in Eureka) or a nicer ID than UUID that can be more descriptive on the UI.

File: sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/TelemetryData.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  /**
+   * The name of this adapter instance. This value must be defined in the configuration file.
+   * Maximum length of adapter name is 64.
+   */
+  @NotBlank()
+  @Length(max = 64)
+  private String adapterName;
+
+  /** The version of the SIP Framework. This value must be defined in the configuration file. */
+  private String version = "1.0.0";
```
Comment (LetoBukarica) [inline sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/TelemetryData.java]: This looks like it should be fetched programmatically.

File: sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/TelemetryData.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+  /**
+   * The interval the adpater registeres at the SIP Backend. The value of register interval has to
+   * between 1000ms and 120000ms. Default value is 30000 milliseconds.
+   */
+  @Min(1000)
+  @Max(120000)
+  private Long interval = 30000L;
```
Comment (LetoBukarica) [inline sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/TelemetryData.java]: Should be fetched from Property file (in the same way as InstanceId and InstanceUri)

File: sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/SIPRegistrationProperties.java (hunk 4 above)
Comment (LetoBukarica) [reply sip-core/src/main/java/de/ikor/sip/foundation/core/premiumsupport/registration/SIPRegistrationProperties.java:41]: We don't have a need for it right now. Instance ID is used only when there is a backend so I'll resolve this for now.

**Comment by vladiIkor** [comment]
Moved to other repo

**Pull request closed without merging**